### ISSUE TITLE
[12] GradNorm: Gradient Normalization for Adaptive Loss Balancing in Deep Multitask Networks

### ISSUE THREAD

**Issue by dhkim0225** [dhkim0225/1day_1paper#39]
[paper](https://arxiv.org/abs/1711.02257)
[youtube-naver_techtalk](https://www.youtube.com/watch?v=klFFAndQia8)

간단하고 잘 동작하는 방법.
Uncertainty 대가 Yarin Gal 이 제시한 방법보다 잘 동작하는 편.

관련해서 이후에 나온 work 들은 다음과 같다. (시간날 때 살펴보자)
[Multi-Task Learning as Multi-Objective Optimization](https://arxiv.org/abs/1810.04650) - NeurIPS18
[Multi-Loss Weighting with Coefficient of Variations](https://openaccess.thecvf.com/content/WACV2021/papers/Groenendijk_Multi-Loss_Weighting_With_Coefficient_of_Variations_WACV_2021_paper.pdf) - WACV21

Cov-weighting (WACV21) 논문을 보면 생각보다 multi-objective (NeurIPS18) 의 성능이 좋지 못하다 ~(task by task)~. broad 하게 널리 쓰이고, 검증된 방법은 grad-norm 이라 판단이 된다.

## GradNorm
![image](https://user-images.githubusercontent.com/16400591/137842850-50aa7853-c1a8-40f5-994a-85d96306669b.png)
~아 ㅋㅋ 오캄 면도날 EZ~

Task 별로 난이도가 다를 텐데 MTL에서는 요게 문제가 된다.
특정 task 는 너무 빨리 수렴해서 overfitting 이 진행되는 것.
이걸 적당히 normalize 시켜서 adaptive 하게 loss balancing 을 수행한다.

이 문제를 해결하기 위해 task 별로 priority를 정해주는 work 도 있다.
[Dynamic-Task-Prioritization-for-Multitask-Learning](https://www.semanticscholar.org/paper/Dynamic-Task-Prioritization-for-Multitask-Learning-Guo-Haque/995495e36f4a2af999875ea4f197ca98c5e5c8de) - ECCV18

## Notation
Loss 는 w_i weight 가 곱해져서 최종 loss를 만들게 된다.
![image](https://user-images.githubusercontent.com/16400591/137844275-97419c6b-2d9a-4e72-b8ad-18592cca8876.png)

w_i weight 는 매 step-by-step t 를 이용해 표기할 수 있다. 
w_i(t) 는 그니까 t 시점의 i번째 task의 weight
![image](https://user-images.githubusercontent.com/16400591/137844462-e69d359d-1b1d-4716-8a40-6c72f6ffc068.png)

hyperparameter α는 원래의 gradient로 되돌리는 복원력의 강도.
α 가 0이면 task-wise grad 가 완전히 같은 크기로 조정된다.
NYU dataset 에서는 1.5 (0~3 모두 성능향상에 도움이 되었음)
toy example 들에서는 0.12 
~근데 이게 좀 찝찝하다. 결국 데이터마다 알파값을 잘 조정해야하는 것 아닌가~

추가적인 notation 들
![image](https://user-images.githubusercontent.com/16400591/137845419-214a15c8-1a8d-4609-bbf1-4ce58f44b2c4.png)
![image](https://user-images.githubusercontent.com/16400591/137845562-f40cbcf5-45a6-4ce7-ae4a-45b350dabed6.png)

## Algorithm & Code
바로 algorithm, 코드를 살펴보자.
notation 만 이해하고 봐도 굉장히 직관적이다.
![image](https://user-images.githubusercontent.com/16400591/137843977-41f6e1ee-0661-4436-8021-a4070ca96843.png)

코드는 unofficial pytorch code (https://github.com/brianlan/pytorch-grad-norm/blob/master/train.py#L93)를 참고해서 정리해 보았다.
```python
# get layer of shared weights
W = model.get_last_shared_layer()

# get the gradient norms for each of the tasks
# G^{(i)}_w(t) 
norms = []
for i in range(len(task_loss)):
    # get the gradient of this task loss with respect to the shared parameters
    gygw = torch.autograd.grad(task_loss[i], W.parameters(), retain_graph=True)
    # compute the norm
    norms.append(torch.norm(torch.mul(model.weights[i], gygw[0])))
    norms = torch.stack(norms)

# compute the inverse training rate r_i(t) 
loss_ratio = task_loss.data.cpu().numpy() / initial_task_loss  # \curl{L}_i 
inverse_train_rate = loss_ratio / np.mean(loss_ratio)  # r_i(t)

# compute the mean norm \tilde{G}_w(t) 
mean_norm = np.mean(norms.data.cpu().numpy())

# compute the GradNorm loss 
# this term has to remain constant
constant_term = torch.tensor(mean_norm * (inverse_train_rate ** args.alpha), requires_grad=False)
constant_term = constant_term.cuda()

# this is the GradNorm loss itself
grad_norm_loss = torch.tensor(torch.sum(torch.abs(norms - constant_term)))

# compute the gradient for the weights
model.weights.grad = torch.autograd.grad(grad_norm_loss, model.weights)[0]
